Core: Restrictive fill in batches to reduce state sweeping costs

Mysteryem opened the pull request:
## What is this fixing or adding?

Reduces the duration of progression fill by about 40% on average. For smaller generations, the size of the reduction can vary.

A large performance cost in `fill_restrictive` is sweeping to create the `maximum_exploration_state` because it sweeps from `base_state` every time.

This patch breaks up `fill_restrictive` to fill in batches of items at a time. Each batch creates its own `batch_base_state`, and additionally a sweep state created from the batch's base state (`_partial_exploration_state`). The advantage to this is that each `maximum_exploration_state` can be created from the batch's `_partial_exploration_state`, which has already reached many regions and locations, rather than having to sweep from `base_state` each time.

As a small bonus, the batch's `_partial_exploration_state` can sometimes also be used to create a `swap_state`, improving swap performance. A successful swap can sometimes force the batch's `_partial_exploration_state` to be re-created because a swap can change the placement of an item that the `_partial_exploration_state` has already collected. Re-creating the `_partial_exploration_state` is deferred as long as possible, but even still, occasionally adding a single extra sweep on a successful swap is negligible when each swap attempt has to sweep to create a new `swap_state` already.

There are no expected changes to the contents of the output of generations with this patch.

## How does this work?

In AP's current fill code, each time it wants to place items, it will:
1. Take items out of the `item_pool` with the intent to place them
2. Start from a copy of the `base_state`
3. Collect all items remaining in the `item_pool` into this new state
4. Sweep this new state, to collect all reachable, already-placed items
5. Use this new state (`maximum_exploration_state`) to correctly place the items, that were taken from the item pool, into locations that are reachable without the items that are being placed

What this PR does, is repeatedly split the `item_pool` into two parts, items that are in the current batch, and are going to be placed, and items that are *not* in the current batch, and won't be placed until a batch later on.

I think the easiest way to think about how this PR changes AP's fill code is to imagine cutting the item pool in half:

With this PR, each time the fill code wants to place items, it will:

1.
   1. Split the `item_pool` into two halves
   2. Start from a copy of the `base_state`
   3. Collect all items in the *first* half of the `item_pool` into this new state
   4. Sweep this new state to collect all reachable, already-placed items
   5. Use this new state (`_partial_exploration_state`) as a starting point for all item placements from the *second* half of the `item_pool`
1.
   1. Take items out of the *second* half of the `item_pool`, with the intent to place them
   2. Start from a copy of the `_partial_exploration_state`
   3. Collect all items remaining in the *second* half of the `item_pool` into this new state
   4. Sweep this new state, to collect all reachable, already-placed items
   5. Use this new state (`maximum_exploration_state`) to correctly place the items, that were taken from the item pool, into locations that are reachable without the items that are being placed
   6. Repeat each of the steps in **2.** until all items from the *second* half of the `item_pool` have been placed, then start back at **1.** again

The creation of the `_partial_exploration_state` costs about as much as going through each of the steps in AP's current fill code, but it is more efficient to create new `maximum_exploration_state`s from this `_partial_exploration_state` than it is to create `maximum_exploration_state`s from `base_state` because the `_partial_exploration_state` has already done a lot of the work that would need to be re-done when creating `maximum_exploration_state`s from `base_state`.

The actual PR code does not split the `item_pool` into two equal halves, but two parts with different sizes, with those sizes chosen to try to reduce the amount of duplicate work that needs to be done. And each batch also creates a `batch_base_state` that only collects items not in the current batch of items to place, `_partial_exploration_state` starts as a copy of `batch_base_state` and is re-created from `batch_base_state` when it has to be destroyed to due a swap.

## How was this tested?

Multiple generations were performed with sets of template yamls. The generation durations before and after were recorded over 5 generations at 84c2d70d9ada and averaged.

Core-verified worlds with template options are known to be deterministic, and the same result was observed before and after given the same seed. Seeds known to require swapping in progression fill have also been compared to be the same before and after.

The scaling of the batch size was tested by adding extra debug output and using a yaml for the unsupported Jak and Daxter AP World with 1-orbsanity enabled because that adds a little over 2000 progression items for that player.

| `python -O .\Generate.py --skip_output --skip_prog_balancing --seed 1` | Original/s | With patch/s | Percent reduction |
| -- | -- | -- | -- |
| 1 of each template yaml from A Hat in Time to Meritous, except Final Fantasy (37 worlds) | 13.7 | 8.0 | 41.4% |
| 1 of each template yaml from Muse Dash to Zillion, except Sudoku (37 worlds) | 26.8 | 13.9 | 48.3% |
| 1 of each template yaml, except Final Fantasy and Sudoku (74 worlds) | 58.4 | 33.4 | 42.8% |
| 2 of each template yaml, except Final Fantasy and Sudoku (148 worlds) | 133.4 | 78.3 | 41.3% |
| 4 of each template yaml, except Final Fantasy and Sudoku (296 worlds) | 303.2 | 208.1 | 31.4% |

Mysteryem committed:
Core: Restrictive fill in batches to reduce state sweeping costs

Reduces the duration of progression fill by about 40% on average.
For smaller generations, the size of the reduction can vary.

A large performance cost in fill_restrictive is sweeping to create the
maximum exploration state because it sweeps from base_state every time.

This patch breaks up fill_restrictive to fill in batches of items at a
time. Each batch creates its own base state, and additionally a sweep
state created from the batch's base state. The advantage to this is that
the maximum exploration state can be created from the batch's sweep
state rather than having to sweep from base_state.

As a small bonus, the batch's sweep state can sometimes also be used to
create a swap_state, increasing swap performance. A successful swap will
sometimes destroy the batch's sweep state and require it to be
re-created, but because swap sweeps to create a swap_state for every
location it checks, occasionally adding a single extra sweep on a
successful swap is negligible.

There are no expected changes to the contents of the output of
generations with this patch.

beauxq commented on Fill.py:
```diff
@@ -62,18 +62,137 @@ def fill_restrictive(multiworld: MultiWorld, base_state: CollectionState, locati
         items_to_place = [items.pop()
                           for items in reachable_items.values() if items]
         for item in items_to_place:
+            for p, batch_pool_item in enumerate(batch_item_pool):
+                if batch_pool_item is item:
+                    batch_item_pool.pop(p)
+                    break
             for p, pool_item in enumerate(item_pool):
                 if pool_item is item:
                     item_pool.pop(p)
                     break
```
I think there's a reason for `item_pool` to be done this way where `batch_item_pool` wouldn't have the same reason.

`item_pool` is an object passed as a parameter to this function that is mutated. Other code might depend on the pointer not changing which object it's pointing to.

`batch_item_pool` is not passed as a parameter to this function, so it's not important to keep the pointer pointing at the same object.
So we could gather up the indexes that need to be removed and then replace `batch_item_pool` with a new list with the gathered indexes filtered out, reducing the time complexity by one exponent.

In my trace, the `batch_item_pool.pop(p)` seemed to always pop index 0, which is the slowest `pop`.
`item_pool.pop(p)` was popping from the end, which is the fastest `pop`.

I didn't measure the performance of the suggested change, so I'm not sure whether it's worth it or even bad.

Mysteryem replied:
I determined it to be irrelevant to performance. I also suspect you ran with only a single player in your generation. I think that if you were to have more than one player, all the items for each player would be added into `batch_item_pool` in a row, so the indices to pop for players after the first would not be at the end of the list.

`item_pool` does not need to be done this way either, it's clearly more efficient to iterate in reverse, but it's even more efficient to replace the entire `item_pool` at the start with a dictionary keyed by the `id()` of each item. But both of these changes are basically irrelevant unless you're generating with many many worlds with both fast access rules and a large percentage of progression items. e.g. generating with 400 default AHiT yamls (~43200 progression items out of ~89974 items total) goes from about 90s to 50s on main (with progression balancing disabled and --skip_output).

Mysteryem committed:
Merge branch 'main' into fill_restrictive_batches

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into fill_restrictive_batches

# Conflicts:
#	Fill.py

Mysteryem committed:
Remove optimisation when running out of items to place

Priority fill will usually have far fewer locations than there are
progression items to place, so this optimisation will rarely have an
effect.

Even when there are fewer items than priority locations, the remaining
items are going to be fewer than the size of the pool, so the
performance saving should be minimal.

Mysteryem committed:
Add comments

Mysteryem committed:
Simplify random sampling

Mysteryem committed:
Add dummy value and typing for `placement_order_iter`

Mysteryem committed:
Re-arrange and rename constants

Mysteryem committed:
Fix distribution of placed items

If there are 2 players, Player1 has 100 items and Player2 has 1 item,
the next item to place should have a 50% chance of belonging to Player1.

The previous attempt at this was much faster, but did not account for
this. This now uses the same code for picking items to place as the
original PR that added `one_item_per_player=False` behaviour.

This is a little considerable slower. 25 template AHiT yamls with
`priority_locations: Everywhere` and no progression balancing goes from
a 6s fill to a 7s fill. This still surpasses the speed on main, which
takes 39s for the same fill.

Mysteryem committed:
Fix unfairness at batch boundaries

If a player still has items to place, but their items in the current
batch has been exhausted, they should still have a chance of having
their item be the next item to place, but, previously, they had no
chance.

Now, all players with items remaining have an equal chance to have their
next item placed next, even if that player has no items remaining in the
batch. If this happens, a new batch is started, and their item is
forcefully placed next.

Mysteryem committed:
Invert if condition in batch creation

It now matches the logical order of the `if one_item_per_player:` if
statement just below it.

Mysteryem commented:
The random picking of the player to place an item added for `one_item_per_player=False` was a bit of a pain because, for fairness, it has to be allowed pick to place an item that is not in the current batch, forcing the current batch to be discarded early and a new batch created. This situation only really happens towards the end of a batch, so it's not a big loss when it does happen.

It looks like it's all working now, though swap could do with some more testing because I have been struggling to generate seeds that actually fall into swap in priority fill (with priority fill changed to use `swap=True`), that recover and successfully generate.

25 Template AHiT yamls with `priority_locations: [Everywhere]` goes from 39.5s on main (1ac8349bd427e7ba2980c4083e28ceaeafa923e8) to 7.5s with this PR, using `python -O .\Generate.py --skip_output --seed 1` and produces identical results. AHiT has a rather high percentage of progression items, so this is not representative of most games.

Mysteryem committed:
Forgot to increase placements remaining when swapping into the batch

It might not even be necessary to track the number of batch placements
remaining when placing 1 item at a time because it will automatically
start the next batch if it tries to place an item that is not in the
batch.

beauxq commented on Fill.py:
```diff
@@ -63,26 +63,174 @@ def fill_restrictive(multiworld: MultiWorld, base_state: CollectionState, locati
+        explore_locations = multiworld.get_filled_locations(item.player) if single_player_placement else None
+
+        # Create the initial batch sweep state or recreate it if it was destroyed by a swap.
+        if batch_sweep_state is None:
```
```suggestion
        if batch_sweep_state is None:
            assert batch_base_state is not None, "should be set when creating the batch"
```
If someone changes the code to break this, this will make the error message more clear.
This cleans up the type checking so we can see other problems more easily.
(and `assert` statements are removed in optimization)

beauxq replied:
Actually, can this be moved outside of the `if` block?
There's another place at line 342 where it's assumed to be set.
Putting it outside of this `if` block would cover both of those.

Mysteryem replied:
Un-placing an item that `batch_sweep_state` has already collected puts the un-placed item back into `batch_item_pool` if there is space in the current batch. (There is space in the current batch if the player the item belongs to did not have enough items remaining in `reachable_items[unplaced_item.player]` to fill out the full number-of-items-per-player of the current batch.)

The issue is that the next `maximum_exploration_state` is created by sweeping from `batch_sweep_state` with the items in `batch_item_pool + unplaced_items` collected into its inventory. This would lead to the un-placed item being in both `batch_sweep_state`'s inventory and being in `batch_item_pool` simultaneously, so that one un-placed item would be collected twice by the next `maximum_exploration_state` for the current batch.

To prevent this, `batch_sweep_state` is destroyed when swapping (`batch_sweep_state = None`), which is why the `if batch_sweep_state is None:` check is necessary.

---

I think I can safely move the `batch_sweep_state = None` into the conditional branch that puts un-placed items back into the current batch, so that `batch_sweep_state` is not destroyed when the un-placed item has to be placed in a later batch.

---

I don't think I can use `batch_sweep_state.remove()` to remove the un-placed item because that will not remove any other items that could only be collected because the un-placed item had been collected.

I think if `batch_sweep_state.remove()` was used, `maximum_exploration_state` would still end up correct to start with, despite sweeping from the invalid `batch_sweep_state`, because `maximum_exploration_state` collects the un-placed item that was put into `batch_item_pool`, but when it is time to re-place the un-placed item, the un-placed item is removed from `batch_item_pool` prior to creating `maximum_exploration_state`, so `maximum_exploration_state` would be incorrect when placing the item.

`batch_sweep_state` is also the preferred state to sweep from to create `swap_state` when swapping. Allowing `batch_sweep_state` to become invalid would prevent this swap optimisation.

beauxq replied:
I haven't been able to wrap my head around all of that yet, but the question I asked (unclearly) is more about `batch_base_state` rather than `batch_sweep_state`
(You didn't mention `batch_base_state` in your reply.)

`batch_base_state` is initialized to `None`, but it looks like it will always be a `CollectionState` (not `None`) before it ever gets referenced.
(And once it changes to `CollectionState`, it never changes to `None` again.)
That's all good if that's the case, and I was kind of asking you to confirm that.

The issue that came to mind is that it's slightly fragile with the initialization to a `CollectionState` being inside an `if` statement (line 115).
With AP code always changing (especially `Fill.py`), I can imagine things getting moved or tweaked by someone in the future so that initialization gets missed.

So outside of that initializing `if`, just before it gets referenced, I'm suggesting an `assert` to make sure `batch_base_state is not None` before it gets referenced.

At first, I only noticed the reference inside this `if` block
```python
        if batch_sweep_state is None:
            # `locations=None` defaults to `multiworld.get_filled_locations()`, so only get it once for both the batch
            # sweep state and the maximum exploration state.
            if explore_locations is None:
                explore_locations = multiworld.get_filled_locations()
            batch_sweep_state = sweep_from_pool(batch_base_state, locations=explore_locations)
```
where it would be a problem if `batch_base_state` were `None`

Because I only first noticed it there, I suggested the `assert` inside that `if` block.
But then I noticed references to `batch_base_state` further down, so the `assert` right before the `if` statement would be before all of the references.

```python
        assert batch_base_state is not None, "should be set when creating the batch"

        # Create the initial batch sweep state or recreate it if it was destroyed by a swap.
        if batch_sweep_state is None:
		    ...
```

Mysteryem replied:
Ah, I failed to properly read what the suggested change was doing.

`batch_base_state` should always be non-None after the first batch has been created.

I needed to assign the `batch_base_state` outside the loop because the assignment that occurs when creating the first batch occurs inside an if statement. Using the `batch_base_state` variable outside that if statement would have PyCharm complaining that `batch_base_state` is potentially being referenced before assignment.

I assigned to `None` rather than a `CollectionState` instance to reduce the likelihood of silent errors occurring and generation crashing further on.

I can add an assertion with a message.

Mysteryem committed:
Only destroy batch sweep state when adding back into the current batch

This can be quite a bit faster for seeds that have to swap. A test seed
went from 8s to 7s.

Mysteryem committed:
Assert that batch_base_state is not None after placements have started

Mysteryem committed:
Refactor batch code into classes

Mysteryem committed:
Comments and cleanup

Mysteryem committed:
Comment and docstring updates

Mysteryem committed:
Fully encapsulated batches in the batcher

Mysteryem committed:
Move the batch classes inside the batcher class

If you don't care about these classes, you can now collapse all of them
in your IDE by collapsing just _RestrictiveFillBatcher.

Mysteryem committed:
Comment updates

Mysteryem committed:
Fully encapsulate reachable_items/item_pool modification within batcher

Mysteryem commented:
`fill_restrictive` feels a bit too big with all of these changes added into it, to me. I'm thinking it might be a good idea to encapsulate most/all of this new code into a class, to keep the `fill_restrictive` function from growing too much in size.

Mysteryem committed:
Make percentage of largest player pool to place a constant

Mysteryem committed:
Update item_pool pop loop order to match #4536

As of #4536, the item_pool is searched in reverse for better
performance.

batch_item_pool currently has each player's items one after another so
it does not benefit from this, but maybe there could be benefit to
constructing it in an order that is more efficient to pop from.
batch_item_pool is usually much smaller than item_pool, so it will be
quicker to iterate in general.

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into fill_restrictive_batches

# Conflicts:
#	Fill.py

Mysteryem committed:
Update docstrings, comments and type hints

Mysteryem committed:
Rename batch_sweep_state -> batch_partial_exploration_state

Mysteryem committed:
Destroy partial exploration state less often when swapping

A swap can 'damage' the partial exploration state by removing an item
from it, such that it could have collected items from locations that
were only reachable with the removed item. However, the maximum
exploration states and swap states swept from the partial exploration
state will collect the removed item because the removed item gets added
back into `batch_item_pool`.

The items that 'damaged' the partial exploration state have to be
tracked because those items will eventually be removed from
`batch_item_pool` so that they can be placed. Once any of these items
are removed from `batch_item_pool`, the partial exploration state must
be re-created because any maximum exploration or swap states swept from
the 'damaged' partial exploration state would also become invalid.

`batch_partial_exploration_state` has been renamed to
`_partial_exploration_state` to indicate is as private, because the
partial exploration state can now be left in a partially invalid state,
so should only be used with additional care.

Mysteryem committed:
Add a comment about sweeping from base state when getting swap state

This makes it clearer that having to fall back to the conditional branch
that results in sweeping from the batch's base state is undesirable due
to its increased performance cost.

Mysteryem committed:
Defer destroying the partial exploration state and avoid .remove()

.remove() often deletes cached accessibility because it is no longer
knowable if the cached accessibility is correct. This incurs a larger
performance cost because the cached accessibility will need to be
redetermined.

Mysteryem committed:
Comment update and rename _MIN_BATCH_SIZE to _MIN_BATCH_ITEMS_PER_PLAYER

Mysteryem committed:
More comments and docstrings

Mysteryem committed:
Add typing.override decorators

The current overridden methods are pretty resilient to hiding bugs if
only the base methods were renamed because one of the methods is
abstract and the other calls the super() method.

Adding the typing.override decorators only makes the code clearer.

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into fill_restrictive_batches

Mysteryem committed:
Revert "Add typing.override decorators"

typing.override requires 3.12, so can't be used yet.

This reverts commit 2b8930787e82150f1d97e7f4c5644c26a8372910.

Mysteryem commented:
I have pushed a large update that encapsulates the code into classes with a fair amount of documentation.

A couple of changes were made to make swapping into the current batch, an item that was already collected, more efficient by deferring the destruction of the `_partial_exploration_state` (previously `batch_sweep_state`) until any of the swapped items are being placed:
Destroy less often by using `CollectionState.remove()`: https://github.com/ArchipelagoMW/Archipelago/pull/3872/commits/4d9eeeb9f56665e2d67d945a05d208b1a9313111
Defer destruction and avoid using `CollectionState.remove()` https://github.com/ArchipelagoMW/Archipelago/pull/3872/commits/8dfaae4263dc8558f7068cfa088b6347e836009a

Mysteryem committed:
More comment updates

Mysteryem committed:
Merge branch 'main' into fill_restrictive_batches

Mysteryem committed:
Also remove from locations_checked when the location becomes unreachable

Deferring partial_exploration_state destruction when swapping into the
current batch also removes the swap location from locations_checked, so
the exceedingly rare conditional branch for swapping into a future batch
when the location was reachable before, but is no longer reachable after
swapping the item at that location, should also remove the swap location
from locations_checked because it was removing the swap location from
advancements.

The comment has been updated to remove mention of minimal accessibility
because this is not a case that happens specifically because of minimal
accessibility.

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into fill_restrictive_batches

# Conflicts:
#	Fill.py

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into fill_restrictive_batches

# Conflicts:
#	Fill.py

Mysteryem committed:
Cleanup: Fix typo in docstring

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into fill_restrictive_batches

Mysteryem commented:
I have retimed the generations again on current main. Notably, https://github.com/ArchipelagoMW/Archipelago/pull/3812 has been merged since the previous timings.

Note that a few worlds have been removed from core and a few new worlds have been added to core.

---

Old table at 824caaffd0bc (meta.yaml used to disable progression balancing), durations averaged over 5 generations.

| `python -O .\Generate.py --skip_output --seed 1` | Original/s | With patch/s | Percent reduction |
| -- | -- | -- | -- |
| 1 of each template yaml from A Hat in Time to Meritous, except Final Fantasy, (37 worlds) | 13.7 | 9.2 | 32.6% |
| 1 of each template yaml from Minecraft to Zillion, except Sudoku, (36 worlds) | 35.3 | 20.2 | 42.7% |
| 1 of each template yaml, except Final Fantasy and Sudoku (73 worlds) | 101.2 | 50.9 | 49.8% |
| 2 of each template yaml, except Final Fantasy and Sudoku (146 worlds) | 302.9 | 183.4 | 39.4% |
| 4 of each template yaml, except Final Fantasy and Sudoku (292 worlds) | 722.0 | 503.7 | 30.2% |

---

New table at ad17c7fd216b, durations averaged over 5 generations.
| `python -O .\Generate.py --skip_output --skip_prog_balancing --seed 1` | Original/s | With patch/s | Percent reduction |
| -- | -- | -- | -- |
| 1 of each template yaml from A Hat in Time to Meritous, except Final Fantasy (37 worlds) | 14.3 | 8.3 | 42.0% |
| 1 of each template yaml from Muse Dash to Zillion, except Sudoku (37 worlds) | 32.0 | 20.0 | 37.4% |
| 1 of each template yaml, except Final Fantasy and Sudoku (74 worlds) | 64.6 | 43.4 | 32.8% |
| 2 of each template yaml, except Final Fantasy and Sudoku (148 worlds) | 153.1 | 102.8 | 32.9% |
| 4 of each template yaml, except Final Fantasy and Sudoku (296 worlds) | 364.0 | 263.6 | 27.6% |

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into fill_restrictive_batches

Mysteryem commented:
I have retimed the generations on current main again, now that https://github.com/ArchipelagoMW/Archipelago/pull/4228 has been merged.

---

Old table at ad17c7fd216b, durations averaged over 5 generations.
| `python -O .\Generate.py --skip_output --skip_prog_balancing --seed 1` | Original/s | With patch/s | Percent reduction |
| -- | -- | -- | -- |
| 1 of each template yaml from A Hat in Time to Meritous, except Final Fantasy (37 worlds) | 14.3 | 8.3 | 42.0% |
| 1 of each template yaml from Muse Dash to Zillion, except Sudoku (37 worlds) | 32.0 | 20.0 | 37.4% |
| 1 of each template yaml, except Final Fantasy and Sudoku (74 worlds) | 64.6 | 43.4 | 32.8% |
| 2 of each template yaml, except Final Fantasy and Sudoku (148 worlds) | 153.1 | 102.8 | 32.9% |
| 4 of each template yaml, except Final Fantasy and Sudoku (296 worlds) | 364.0 | 263.6 | 27.6% |

---

New table at 84c2d70d9ada, durations averaged over 5 generations.
| `python -O .\Generate.py --skip_output --skip_prog_balancing --seed 1` | Original/s | With patch/s | Percent reduction |
| -- | -- | -- | -- |
| 1 of each template yaml from A Hat in Time to Meritous, except Final Fantasy (37 worlds) | 13.7 | 8.0 | 41.4% |
| 1 of each template yaml from Muse Dash to Zillion, except Sudoku (37 worlds) | 26.8 | 13.9 | 48.3% |
| 1 of each template yaml, except Final Fantasy and Sudoku (74 worlds) | 58.4 | 33.4 | 42.8% |
| 2 of each template yaml, except Final Fantasy and Sudoku (148 worlds) | 133.4 | 78.3 | 41.3% |
| 4 of each template yaml, except Final Fantasy and Sudoku (296 worlds) | 303.2 | 208.1 | 31.4% |

darkfire006 reviewed (approved):
I have used this to generate a sync quickly and effectively every week!

gerbiljames reviewed (approved):
Tested by generating 1000 random YAMLs for Pokemon Crystal, no failures. Reduced the time for this by ~6%, which may be within the margin of error.

This has also been used to generate the community syncs hosted by 360Chrism, it's very much battle tested. LGTM.
